src/DataGrid.tsx

```diff
@@ -617,29 +597,44 @@ function DataGrid<R, SR, K extends Key>(
     updateRow(columns[selectedPosition.idx], selectedPosition.rowIdx, selectedPosition.row);
   }
 
-  function handleCopy() {
+  function handleCopy(event: React.ClipboardEvent<HTMLDivElement>) {
+    if ((!onPaste && !onCopy) || !selectedCellIsWithinViewportBounds) return;
     const { idx, rowIdx } = selectedPosition;
     const sourceRow = rows[rowIdx];
     const sourceColumnKey = columns[idx].key;
+    const cellEvent = createCellEvent(event);
+    onCopy?.({ sourceRow, sourceColumnKey }, cellEvent);
+    if (cellEvent.isGridDefaultPrevented()) return;
+
     setCopiedCell({ row: sourceRow, columnKey: sourceColumnKey });
```

Comment:
if developers use `event.preventGridDefault()` in `onCopy` handler, the default behavior to set the copied cell internally would be prevented.